modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java

```diff
@@ -0,0 +1,166 @@
+     * @param baselineNodes Nodes in baseline.
+     * @param partitions Number of partitions in a table.
+     * @param replicas Number of replicas for a table.
+     * @param revision Revision of Meta Storage that is specific for the assignment update.
+     * @param metaStorageMgr Meta Storage manager.
+     * @return Future representing result of updating keys in {@code metaStorageMgr}
+     */
+    public static @NotNull CompletableFuture<StatementResult> updatePendingAssignmentsKeys(
+            String partId, Collection<ClusterNode> baselineNodes,
+            int partitions, int replicas, long revision, MetaStorageManager metaStorageMgr, int partNum) {
+        ByteArray partChangeTriggerKey = partChangeTriggerKey(partId);
+
+        ByteArray partAssignmentsPendingKey = partAssignmentsPendingKey(partId);
+
+        ByteArray partAssignmentsPlannedKey = partAssignmentsPlannedKey(partId);
+
+        ByteArray partAssignmentsStableKey = partAssignmentsStableKey(partId);
+
+        byte[] partAssignmentsBytes = ByteUtils.toBytes(
+                AffinityUtils.calculateAssignments(baselineNodes, partitions, replicas).get(partNum));
+
+        //    if empty(partition.change.trigger.revision) || partition.change.trigger.revision < event.revision:
+        //        if empty(partition.assignments.pending) && partition.assignments.stable != calcPartAssighments():
+        //            partition.assignments.pending = calcPartAssignments()
+        //            partition.change.trigger.revision = event.revision
+        //        else:
+        //            if partition.assignments.pending != calcPartAssignments
+        //                partition.assignments.planned = calcPartAssignments()
+        //                partition.change.trigger.revision = event.revision
+        //            else
+        //                remove(partition.assignments.planned)
+        //    else:
+        //        skip
+        var iif = If.iif(or(exists(partChangeTriggerKey), value(partChangeTriggerKey).lt(ByteUtils.longToBytes(revision))),
+                If.iif(and(notExists(partAssignmentsPendingKey), value(partAssignmentsStableKey).ne(partAssignmentsBytes)),
+                        ops(
+                                put(partAssignmentsPendingKey, partAssignmentsBytes),
+                                put(partChangeTriggerKey, ByteUtils.longToBytes(revision))
+                        ).yield(),
+                        If.iif(value(partAssignmentsPendingKey).ne(partAssignmentsBytes),
+                                ops(
+                                        put(partAssignmentsPlannedKey, partAssignmentsBytes),
+                                        put(partChangeTriggerKey, ByteUtils.longToBytes(revision))
+                                ).yield(),
+                                ops(remove(partAssignmentsPlannedKey)).yield())),
+                ops().yield());
+
+        return metaStorageMgr.invoke(iif);
```

Comment:
Let's also wrap it with busy lock.

Reply:
Calls of this util methods already covered by TableManager busy lock.

Reply:
Agreed.